File: day06/aoc2025_06.scala
import scala.collection.mutable.ListBuffer
object aoc2025_06 {
    def do_math(nums: Iterable[String], op: Char): Long = {
        op match 
            case '+' => nums.map(_.strip().toLong).sum
            case '*' => nums.map(_.strip().toLong).product
    }
    def main(args: Array[String]): Unit = {
        val lines = scala.io.Source.fromFile("day06/input06").getLines.toArray
        // lines.foreach(x => println(x.mkString.take(150)))
        val matrix: Array[Array[String]] = lines.map(_.trim.split("\\s+"))
        val num_cols = matrix(1).length 

        val part1 = Range(0, num_cols).map { i =>
                val col = matrix.map(_(i))
                do_math(nums = col.dropRight(1), op = col.last(0))
            }.sum  
        println("Part 1: " + part1)

        var part2 = 0L
        var list: ListBuffer[String] = ListBuffer()
        for (i <- Range(lines(0).length -1, -1, -1)) {
            val nums: Array[Char] = lines.map(_(i))
            if (!nums.forall(_ == ' ')) {
                list += nums.dropRight(1).mkString
                if ("+*".contains(nums.last)) {
                    part2 += do_math(nums = list, nums.last)
                    list = ListBuffer()
                }
            }
        }
        println("Part 2: " + part2)
    }
}

File: day06/aoc2025_06.ts
import { readFileSync } from "node:fs";

function do_math(nums: string[], op: string): number {
    switch (op) {
        case '+':
            return nums.map(x => x.trim()).map(Number).reduce((x, y) => x + y)
        case '*':
            return nums.map(x => x.trim()).map(Number).reduce((x, y) => x * y)
        default:
            throw new Error(`Unknown op: ${op}`)
    }
}

const lines: string[] = readFileSync("day06/input06", "utf-8")
    .split("\n")
    .slice(0, -1)

const matrix: string[][] = lines.map(x => x.trim().split(/\s+/))
const num_cols = matrix[0].length
let part1 = 0
for (let i = 0; i < num_cols; i++)
    part1 += do_math(matrix.slice(0, -1).map(x => x[i]), matrix[matrix.length - 1][i])
console.log('Part 1:', part1)

let part2 = 0
let stack: string[] = []
for (let i = lines[0].length - 1; i >= 0; i--) {
    const nums = lines.map(x => x[i])
    if (nums.every(c => c === ' ')) continue
    stack.push(nums.slice(0, -1).join(''))
    if ('+*'.includes(nums[nums.length - 1])) {
        part2 += do_math(stack, nums[nums.length - 1])
        stack = []
    }
}
console.log('Part 2:', part2)
